test/spark.js

```diff
@@ -0,0 +1,70 @@
+import Spark from '../lib/spark.js'
+import { test } from 'zinnia:test'
+import { assertInstanceOf, assertEquals } from 'zinnia:assert'
+
+test('getRetrieval', async () => {
+  const retrieval = { retrieval: 'retrieval' }
+  const fetch = async (url, opts) => {
+    assertEquals(url, 'https://spark.fly.dev/retrievals')
+    assertEquals(opts.method, 'POST')
+    return {
+      async json () {
+        return retrieval
+      }
+    }
+  }
+  const spark = new Spark({ fetch })
+  assertEquals(await spark.getRetrieval(), retrieval)
+})
+
+// TODO: test more cases
+test('fetchCAR', async () => {
+  const URL = 'url'
+  const fetch = async url => {
+    assertEquals(url, URL)
+    return {
+      status: 200,
+      ok: true,
+      body: {
+        [Symbol.asyncIterator]: () => {
+          let i = 0
+          return {
+            async next () {
+              if (i === 0) {
+                i++
+                return {
+                  value: new Uint8Array([1, 2, 3]),
+                  done: false
+                }
+              } else {
+                return {
+                  value: undefined,
+                  done: true
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+  const spark = new Spark({ fetch })
+  const stats = await spark.fetchCAR(URL)
+  assertInstanceOf(stats.start, Date)
+  assertInstanceOf(stats.firstByte, Date)
+  assertInstanceOf(stats.end, Date)
+  assertEquals(stats.byteLength, 3)
+  assertEquals(stats.status, 200)
+})
+
+test('submitRetrieval', async () => {
+  const fetch = async (url, opts) => {
+    assertEquals(url, 'https://spark.fly.dev/retrievals/0')
+    assertEquals(opts.method, 'PATCH')
+    assertEquals(JSON.parse(opts.body), { success: true })
+    assertEquals(opts.headers['Content-Type'], 'application/json')
+    return { status: 200 }
+  }
+  const spark = new Spark({ fetch })
+  await spark.submitRetrieval(0, { success: true })
```

Comment:
I think this test will not detect when `submitRetrieval` does not call `fetch` at all or calls it multiple times.

Fix: in the stub, record arguments into an array of invocations. Later, assert that this array has the correct number of items with the expected properties.

```js
const requests = []

const fetch = async (url, opts) => {
  requests.push({
    url, 
    method: opts.method, 
    contentType: opts.headers['Content-Type'], 
    body: opts.body
  })
  return { status: 200 }
}

//...

assertEquals(
  requests.map(pick("url", "method", "contentType")),
  [{ url: 'https://spark.fly.dev/retrievals/0', method: 'POST', contentType: 'application/json'}]
)

assertEquals(JSON.parse(requests[0].body), { success: true })
```

When you write the test this way, and something goes wrong, the assertion error message will include many useful details.
- How many requests were made - too few? too many?
- If too many requests were sent, what were their URL, method, and other details?

We could also use something like `Sinn.stub` instead of writing our own stub function recording calls and arguments. In this particular case, I think we get better DX with our custom implementation, though, as we can save only the relevant fields from the Fetch request options.

Reply:
I agree we should not aim at full test coverage. I consider coverage over 80% good enough, and I don't consider 80% coverage as a part of the walking skeleton.

However, I think this is a different case. This test is written in a way that does not make it immediately apparent that it passes when `fetch` was not called. I consider this as a subtle bug in the test, and IMO we should not introduce known bugs into our production-grade code, even if it's test code.

I consider it a good practice to write tests in such a way that they don't accidentally pass for incorrect code and that they report helpful error messages on failures.

The solution I offered solves all of these concerns and is very generic - it can be reused in many different test cases.
- It detects the case when `fetch` was not called or called multiple times
- The error message includes details that allow us to immediately see what went wrong, for all different failure modes (fetch not called, fetch called twice, fetch called with wrong arguments).

Anyhow, if you strongly feel that we should not test the case when fetch was called incorrect number of times, then please try to find a way how to make this aspect obvious to readers.

Reply:
> However, I think this is a different case. This test is written in a way that does not make it immediately apparent that it passes when fetch was not called.

Passing when `fetch` was not called isn't part of the test though. The test isn't claiming that it can pass if `fetch` wasn't called. I wouldn't consider this a bug, it's more that the test isn't as strict or as "good" as it could be.

I have implemented your suggestion, as you feel strongly about this. I personally would have added the suggested logic once we see that the test is insufficient.

Reply:
You're 100% right that the test is better if it verifies that `fetch` was called. To me this is not part of my standard testing criterium. I know this has a potential of bugs, but it's good enough for me. I think we need to work on establishing testing principles that we both can agree on. Again, I would have added this logic once we have a bug that can be traced back to this test, or if any actual uncertainty arose from it.